hamilton/function_modifiers/recursive.py

```diff
@@ -626,3 +633,234 @@ def prune_nodes(nodes: List[node.Node], select: Optional[List[str]] = None) -> L
+                "feed the right data into the right columns."
+            )
+
+        self.initial_schema = columns_to_pass
+        self.dataframe_subdag_param = pass_dataframe_as
+        self.namespace = namespace
+        self.config_required = config_required
+
+    def required_config(self) -> List[str]:
+        return self.config_required
+
+    def _create_column_nodes(
+        self, inject_parameter: str, params: Dict[str, Type[Type]]
+    ) -> List[node.Node]:
+        output_type = params[inject_parameter]
+
+        if self.is_async:
+
+            async def temp_fn(**kwargs) -> Any:
+                return kwargs[inject_parameter]
+        else:
+
+            def temp_fn(**kwargs) -> Any:
+                return kwargs[inject_parameter]
+
+        # We recreate the df node to use extract columns
+        temp_node = node.Node(
+            name=inject_parameter,
+            typ=output_type,
+            callabl=temp_fn,
+            input_types={inject_parameter: output_type},
+        )
+
+        extract_columns_decorator = extract_columns(*self.initial_schema)
+
+        out_nodes = extract_columns_decorator.transform_node(temp_node, config={}, fn=temp_fn)
+        return out_nodes[1:]
+
+    def _get_inital_nodes(
+        self, fn: Callable, params: Dict[str, Type[Type]]
+    ) -> Tuple[str, Collection[node.Node]]:
+        """Selects the correct dataframe and optionally extracts out columns."""
+        initial_nodes = []
+        sig = inspect.signature(fn)
+        input_types = typing.get_type_hints(fn)
+
+        if self.dataframe_subdag_param is not None:
+            inject_parameter = self.dataframe_subdag_param
+        else:
+            # If we don't have a specified dataframe we assume it's the first argument
+            inject_parameter = list(sig.parameters.values())[0].name
+
+        if inject_parameter not in params:
+            raise base.InvalidDecoratorException(
+                f"Function: {fn.__name__} does not have the parameter {inject_parameter} as a dependency. "
+                f"@with_columns requires the parameter names to match the function parameters. "
+                f"If you wish do not wish to use the first argument, please use `pass_dataframe_as` option. "
+                f"It might not be compatible with some other decorators."
+            )
+
+        if input_types[inject_parameter] not in self.allowed_dataframe_types:
+            raise ValueError(f"Dataframe has to be a {self.allowed_dataframe_types} DataFrame.")
+        else:
+            self.dataframe_type = input_types[inject_parameter]
+
+        initial_nodes = (
+            []
+            if self.dataframe_subdag_param is not None
+            else self._create_column_nodes(inject_parameter=inject_parameter, params=params)
+        )
+
+        return inject_parameter, initial_nodes
+
+    @abc.abstractmethod
+    def create_merge_node(self, upstream_node: str, node_name: str) -> node.Node:
```

Comment:
Not sure this is appropriate here, the `with_columns` call could be somewhere in the middle of the DAG and the user doesn't want to collect a LazyFrame just yet for example. 

This node is mostly meant to implement the different `with_columns` flavours -- for example in Polars its straightforward API matching since a `with_columns` method already exists, Pandas we needed to manually implement it ourselves.

Reply:
There's also a [hamilton.plugins.h_polars_lazyframe](https://github.com/DAGWorks-Inc/hamilton/blob/3d98be3e2cf8b8a0e399be77b2d2a7199fe63105/hamilton/plugins/h_polars_lazyframe.py) with a result builder for lazy frames.

From my understanding, the `create_merge_node()` serves the same purpose as `ResultBuilder.build_result()`, which we have implementations for pandas, polars, polars lazy frame, vaex, ibis, and more (IIRC). Wanted to point it out to save you development time. I think there are two paths:

- duplicate code: copy-paste logic from ResultBuilder into this `create_merge_node` method, but make sure to leave a `# TODO` noting that this is the same logic as result builders
- refactor to extract logic: the `.build_result()` or `create_merge_node()` should eventually end up as a method under `hamilton.registry` with `@singledispatch`. Then, the same implementation could be called by `@with_columns` and `ResultBuilder`

Reply:
Not quite. The `ResultBuilder` will take all the end nodes (`final_vars`) and try to build a dataframe out of it. There are two cases and an edge case:

1. If the end node is a dataframe, it adopts that.
2. If the end nodes are a list of series, it combines them into a dataframe.
3. Edge case: In the case the end nodes are mixed the result is funky and definetly not something useful. 

In our case we want to keep the original dataframe and append columns to it, so we need to insert all original columns + new columns into the `ResultsBuilder`. In the case a selected column already exists in the dataframe we need to prefer the selected over the existing. 

Now, when the user initiates the `with_columns` decorator he passes in a subset of the existing columns / will most likely extract himself a subset of existing columns and not all. So, for us to generate the whole frame we will need to essentially extract all the columns in the background to be able to use the `ResultBuilder` logic. 
 
It can be done, but I find the amount of gluing necessary to get what we need a bit much; I find we need a better reason to justify it than just code reusability. Mind you, we will still need to register the `merge_node` in the `registry` anyways and there we can levarage  library-specific `with_columns` implementations / implementing it for libraries that don't have it (e.g. Pandas -- literally one line of code) in a more succinct way.

So all in all, I am not a fan of using `ResultBuilder` here unless there is a good argument about the overall design structure.

Reply:
Per my initial comment, code duplication is totally reasonable and desirable here, but we should add clear comments in `@with_columns` source code AND in the docstrings, docs, and user guides

Code reuse is not a goal in itself. Rather, it's about limiting the number of "concepts/abstraction" we introduce in Hamilton, which impacts downstream maintenance, documentation, user onboarding, etc. A frequent challenge mentioned with Hamilton is that 5 ways to do a single thing are presented to users, which can lead to heterogeneous coding practices within a team.

You highlighted good points that the Reference docs highlight:

- `ResultBuilder().build_result()`:  merge columns and scalars into a dataframe
- `@with_columns` merge operation: subdag that assigns new columns to a dataframe, e.g., merge a dataframe and columns

But then likely scenarios:
> User A has code using `ResultBuilder`. Workflow gets complex, they want to use `@subdag`. I tell them you can't pass a result builder to a `@subdag`. Then,  `@with_columns` should solve the problem, but it doesn't have the same behavior for handling scalars (I'm not fond of it personally)

The main challenge to unify both is that `ResultBuilder` need backwards compatibility for handling scalar. However, as you're writing logic to handle merging of dataframes + columns, it seems a reasonable TODO to update ResultBuilder to handle dataframes + columns + scalars in a backwards compatible (or non-backwards in a future major release).

Then, we have a much simpler story for users "result builder merges X, Y, Z, and `@with_columns` is `@subdag + ResultBuilder`" without other caveats. Part of the documentation effort is also orienting users, making informative suggestion and comparisons, and writing in red bold letters what problem it solves / when it works / when it won't work.

Reply:
See comment above -- I think we don't have the right general pattern yet, and unifying is more trouble than it's worth. Bringing in the `ResultBuilder` construct is, IMO, a little more complexity than needed.

Looking at the bigger picture, we *might* want the next push to be a more generic `with_columns` -- @jernejfrank and I initially discussed that then downscoped, in favor of individual versions that we can merge.

So I think we implment pandas (done), and polars (done), then maybe one more (narwhals, which also gives other stuff for free) or maybe we have enough information to then unify/delete code.

Reply:
I had a call with Jernej in the meantime yesterday. The main notes:
- my intent with this particular thread was to point that out that currently `@with_columns != @subdag + ResultBuilder`. My impression from the docs and the code is that it *should* be the end goal
- the current implementations work, are useful, and should be merged. We just have to add mentions in docs and `# TODO` where appropriate
- the current `@with_columns` doesn't fill scalars like the `ResultBuilder`. But also, Polars `.with_columns()` does fill scalars, but PySpark `.withColumns()` doesn't (see below)

```python
import polars as pl

df = pl.DataFrame({"a": [10, 11], "b": ["aa", "bb"]})
df.with_columns(1)
```

</small><table border="1" class="dataframe"><thead><tr><th>a</th><th>b</th><th>literal</th></tr><tr><td>i64</td><td>str</td><td>i32</td></tr></thead><tbody><tr><td>10</td><td>&quot;aa&quot;</td><td>1</td></tr><tr><td>11</td><td>&quot;bb&quot;</td><td>1</td></tr></tbody></table></div>

```python
import pyspark
spark = pyspark.sql.SparkSession.builder.getOrCreate()
df = spark.createDataFrame([(10, 11), ("aa", "bb")], schema=["a", "b"])
df.withColumns({"c": 1}).show()
```
>[63](https://vscode-remote+wsl-002bubuntu-002d20-002e04.vscode-resource.vscode-cdn.net/home/tjean/projects/dagworks/hamilton/~/projects/dagworks/hamilton/.venv/lib/python3.11/site-packages/pyspark/sql/column.py:63)     jcol = _create_column_from_name(col)
     [64](https://vscode-remote+wsl-002bubuntu-002d20-002e04.vscode-resource.vscode-cdn.net/home/tjean/projects/dagworks/hamilton/~/projects/dagworks/hamilton/.venv/lib/python3.11/site-packages/pyspark/sql/column.py:64) else:
---> [65](https://vscode-remote+wsl-002bubuntu-002d20-002e04.vscode-resource.vscode-cdn.net/home/tjean/projects/dagworks/hamilton/~/projects/dagworks/hamilton/.venv/lib/python3.11/site-packages/pyspark/sql/column.py:65)     raise PySparkTypeError(
     [66](https://vscode-remote+wsl-002bubuntu-002d20-002e04.vscode-resource.vscode-cdn.net/home/tjean/projects/dagworks/hamilton/~/projects/dagworks/hamilton/.venv/lib/python3.11/site-packages/pyspark/sql/column.py:66)         error_class="NOT_COLUMN_OR_STR",
     [67](https://vscode-remote+wsl-002bubuntu-002d20-002e04.vscode-resource.vscode-cdn.net/home/tjean/projects/dagworks/hamilton/~/projects/dagworks/hamilton/.venv/lib/python3.11/site-packages/pyspark/sql/column.py:67)         message_parameters={"arg_name": "col", "arg_type": type(col).__name__},
     [68](https://vscode-remote+wsl-002bubuntu-002d20-002e04.vscode-resource.vscode-cdn.net/home/tjean/projects/dagworks/hamilton/~/projects/dagworks/hamilton/.venv/lib/python3.11/site-packages/pyspark/sql/column.py:68)     )
     [69](https://vscode-remote+wsl-002bubuntu-002d20-002e04.vscode-resource.vscode-cdn.net/home/tjean/projects/dagworks/hamilton/~/projects/dagworks/hamilton/.venv/lib/python3.11/site-packages/pyspark/sql/column.py:69) return jcol
> PySparkTypeError: [NOT_COLUMN_OR_STR] Argument `col` should be a Column or str, got int.